WIP: Cli by click

vlcinsky opened the pull request:
This is still WIP.

# What is done:
- created `nameko.click_cli` as alternative to `nameko.cli`
- added command `cnameko` to allow comparing original and the new command
- help for `cnameko` shown without crash (fixes #618)
- implemented `run`, `show-config`, `shell`
- all `test/cli` tests copied as `test/click_cli` and updated. All are working.

# What is not yet done:
- implementation of `backdoor` subcommand (shall be easy)
- making other than `CLI` related tests completely running (the suite is huge and I am not sure, it was really working completely before my change)

# Interesting side effects:
- some CLI related components refactored to allow re-use out of default CLI. All are in `nameko.click_cli.utils`
  - `setup_config`: now one can use yaml config files incl. env variable rendering out of CLI.
  - `import_service`: renamed to `import_services`. Having module name (opt. with service name) one can use one call to get list of loaded services. CLI related exceptions generalized.
  - `paramtypes` serve click-related parameter/argument types which can be reused in other click based tools.

# Possible side effects in future:
- allow configuring some arguments using env variables and whatever other reasons can be found at https://click.palletsprojects.com/en/7.x/why/

# Next actions:
- complete `backdoor` implementation
- final code review scoped to CLI stuff
- another WIP pull request (still preserving `nameko.cli` and `nameko.click_cli` to allow easy comparison).
- process feedback
- remove old argparse based stuff and keep only the click one.
- pull request

n.b.: I am really impressed by depth of the test suite and what cases it really handles. Great work @iky, @davidszotten , @mattbennett and couple of others.

mattbennett commented:
This is looking great @vlcinsky!

vlcinsky commented:
Done so far:
- [x] created nameko.click_cli as alternative to nameko.cli
- [x] added command cnameko to allow comparing original and the new command
- [x] help for cnameko shown without crash (fixes #618)
- [x] implemented run, show-config, shell
- [x] remove backdoor subcommand, add info how to connect to backdoor to `nameko run` (fixes #625)
- [x] all test/cli tests copied as test/click_cli and updated. All are working.
- [x] refactor `import_services`, `setup_config` and `interact` to nameko.click_cli.utils to make them reusable
- [x] refactor click related reusable things (custom types and option and argument decorators) to nameko.click_cli.myclick (for possible reuse)
- [x] internal code review
- [x] ensure, all test/click_cli tests ares till working (ignoring other ones)

Expected actions:
- [ ] your review of the code (@mattbennett?)
- [ ] accept requests for changes (possibly few iterations)
- [ ] remove old `nameko.cli` stuff, replace with current `nameko.click_cli`
- [ ] (if agreed) squash and rebase
- [ ] final pull request

mattbennett commented on test/click_cli/test_cmain.py:
```diff
@@ -0,0 +1,499 @@
+def expected():
+    define = {"AMQP_URI": "pyamqp://someuser:*****@somehost/"}
+    services = import_services("test.sample:Service")
+    return ExpectedCall(define=define, services=services)
```
I find this hard to follow.

I can only find the `expected` fixture and `ExpectedCall` helper being used once, in `test_run` below. It may be slightly more verbose, but perhaps constructing the `call` objects explicitly inline would be clearer.

vlcinsky replied:
OK. I will do that.

mattbennett commented on test/click_cli/test_cmain.py:
```diff
@@ -0,0 +1,499 @@
+            assert setup_config.call_count == 1
+            assert main_run.call_count == 1
+            assert setup_config.call_args == expected.setup_config_call
+            assert main_run.call_args == expected.main_run_call
```
When I want to assert exactly one call _and_ check the arguments I usually use `assert call_args_list == [call(<arguments>)]` May help make this a bit more terse.

vlcinsky replied:
Great trick. `nameko` test suite is one big lesson on testing with mockups for me.
I will change that.

mattbennett commented on test/click_cli/test_cshow_config.py:
```diff
@@ -0,0 +1,48 @@
+
+
+@pytest.fixture
+def config_file(tmpdir):
```
Doesn't look like this is used.

vlcinsky replied:
Correct. Again comes from the original test suite.

I will either remove the extraneous fixture or use it to shorten the test case itself.I guess you prefer removing the fixture.

mattbennett commented on nameko/click_cli/utils/import_services.py:
```diff
@@ -0,0 +1,110 @@
+        sys.path.insert(0, ".")
+    services = import_services("my_service")  # load my_service.py
+
+See also NamekoModuleServicesParamType if yo want to use it with click.
```
s/yo/you

vlcinsky replied:
yes

vlcinsky commented on nameko/click_cli/myclick/paramtypes.py:
```diff
@@ -0,0 +1,81 @@
+            args = {"value": value, "msg": str(exc)}
+            self.fail(templ.format(**args), param, ctx)
+        except ImportError:
+            # Can we do better error message?
```
The message comes from original code and I preserved it there. I have no problem to remove that comment.

vlcinsky commented on nameko/click_cli/myclick/paramtypes.py:
```diff
@@ -0,0 +1,81 @@
+        """extra_sys_paths: list of paths, which are added to sys.path
+        before iporting services.
+
+        Note: syspath stays modified (side effect)
```
To prevent this side effect, I can modify the sys.path only for the time `import_service` is called and restore it then.

Anyway, what if the code in the imported module attempts to import another module (doing that within running code, not in top level import section). If the lately imported module stays within sys.path, the code can fail.

I think, if we import the modules using modified sys.path, we shall keep the sys.path modified.

What do you think?

mattbennett replied:
Ah, quite right. It's correct that the path remains modified to allow imports at run-time.

I would remove the `(side effect)` from the comment here, because it implies it is unintentional.  It's more correct to say that we are intentionally modifying the path until the process terminates.

vlcinsky replied:
done.

vlcinsky commented on nameko/click_cli/shell.py:
```diff
@@ -0,0 +1,95 @@
+    ctx["n"] = make_nameko_helper()
+
+    runner = ShellRunner(banner, ctx)
+    # import pdb; pdb.set_trace()
```
I will remove that.

vlcinsky commented on nameko/click_cli/myclick/argsopts.py:
```diff
@@ -0,0 +1,68 @@
+"""Custom click option and argument decorators
```
`myclick` is not an convention. In fact, typical click based code does not move that stuff often out of the main module, but here I was thinking of possible reuse (I plan to do one myself for our project possibly cloning the CLI with modified logging configuration).

Searching for proper name `click` was a candidate, which sounded too similar to original click. If you have better idea, I will follow that. `customclick`?

If you think, options with arguments shall share the same module with click types, let me know and I will do it that way.

mattbennett replied:
How about the following:

* `nameko/[click_]cli/click_arguments.py`
* `nameko/[click_]cli/click_options.py`
* `nameko/[click_]cli/click_paramtypes.py`

(Adding `[click_]` here because I'm assuming we will rename this package to simply `cli` when the old one is replaced)

This seems very clear and explicit to me.

vlcinsky replied:
I will do it according to your proposal.

As we are almost done with all comments to this PR, I would now remove the original argparse based code and do the renaming at once.

vlcinsky replied:
Modules renamed.

additionally, I renamed `run.py` to `do_run.py` and `shell.py` to `do_shell.py`.

Remaining steps is now to remove the argparse implementation and rename click_cli to cli

vlcinsky commented on nameko/click_cli/utils/import_services.py:
```diff
@@ -0,0 +1,110 @@
+    """
+    parts = module_name.split(":", 1)
+    if len(parts) == 1:
+        module_name, obj = module_name, None
```
Correct. `obj` comes from the original code. `service_name` sounds better.

vlcinsky commented on nameko/click_cli/utils/import_services.py:
```diff
@@ -0,0 +1,110 @@
+                "Failed to find service, did you mean '{}'?".format(
+                    module_name[:-3].replace("/", ".")
+                )
+            )
```
Good idea, I would like such improvement.

Anyway, i would prefer to do it via separate issue/PR to close this one soon.

mattbennett replied:
Fantastic, thank you. Agree with your suggestion to do this in a separate PR :+1:

vlcinsky replied:
I filed an issue #629

iky replied:
:+1: this would be great extra improvement .)

vlcinsky commented on nameko/click_cli/utils/import_services.py:
```diff
@@ -0,0 +1,110 @@
+                )
+            )
+
+        if not isinstance(service_cls, type):
```
The whole `import_services` is exact copy of the original one at https://github.com/nameko/nameko/blob/v3.0.0-rc/nameko/cli/run.py#L37

I only changed name of the function.

I would propose to keep it as it is now (possibly with an error) and file another issue resolving this magic (which I did not dive into yet).

mattbennett replied:
Ah, sorry! Yes, I agree -- we should save improvements to this module for a later PR.

vlcinsky replied:
I fixed the tests of an object being a class. This fix does not require new PR.

vlcinsky commented:
@mattbennett 7 comments are resolved now.

Remaining 4 comments are await your further reaction.

vlcinsky commented:
Finally I renamed all `click_cli` to `cli`.

There are few things remaining, which I would like to resolve independently:
- updated documentation (e.g. subcommand backport was removed)
- test backport: it used to be tested as part of CLI test suite, but this is now removed. Testing backport is a bit complex as it requires external commands such as `telnet`, `netcat` etc.

After this PR is accepted, I can file an issue related to the docs and testing backport.

Regarding squashing: are you going to do that? I am fine if you do that (or if you keep all the commits even though there are a bit too many).

mattbennett commented on .isort.cfg:
```diff
@@ -1,6 +1,6 @@
 [settings]
 line_length=79
-skip=nameko/cli/run.py,nameko/click_cli/run.py
+skip=nameko/cli/do_run.py
```
Why rename this module from `run` to `do_run`?

vlcinsky replied:
The intention was to keep modules related to actions of subcommand clearly different. I reverted the change so it is named `run.py` and `shell.py` as before.

mattbennett commented:
Thank you @vlcinsky. I'm happy for the docs and backport test to be added in a separate PR.

I have just one remaining question about the `do_run` module name, otherwise all looks good.

I will squash the commits when it's merged.

vlcinsky commented:
Regarding `do_run` module name - I understand you want to name it simply `run`. I will do that.

vlcinsky commented:
All checks have passed!

@mattbennett let me know, if there is anything I am supposed to do now.

davidszotten commented:
the 3.0.0-rc branch is steadily growing.  is it waiting for anything in particular?

mattbennett commented:
@davidszotten https://github.com/nameko/nameko/pull/628 is still open and adds some more backwards compatibility with 2.x, but otherwise I think development is now done.

The docs ought to get an update because there is quite a lot of new functionality here. We wanted to include guide for upgrading to Nameko 3 in there too -- for both service developers and extension developers.

@iky is there anything left other than docs?

iky commented:
@mattbennett yes, only the upgrade guide needed. The pre-release code is used on production on a number of places .)

iky commented:
@vlcinsky any chance meeting on https://cz.pycon.org/2019/ today or tomorrow?

vlcinsky commented:
@iky you are right, it is very close to where I live (25 minutes drive) but I am completing delivery of large project (doc, source code, docker images, python packages incl. some using nameko) due by this Sunday midnight.

I was planning to attend but I gave up. However, if you meet Eliška Vlčinská there, then she is my daughter who got pycon ticket as reward for her excellent graduation.

Feel free to call me or give my daughter your phone number and we can at least have a talk over phone (mine is +420-608979040).
